### ISSUE TITLE
Document or expose a uncurry-like operation

### ISSUE THREAD

**Issue by rpglover64** [nikita-volkov/hasql#59]
After bashing my head against the `Contravariant`, `Divisible`, and `Decidable` interfaces, I have failed to come up with the combinator I want.

I would like a function `uncurry :: (a -> Params b) -> Params (a, b)`. I believe that it is possible to implement its counterpart

```
curry :: Params (a, b) -> a -> Params b
curry p x = contramap (x,) p
```

It should be possible to implement `uncurry` because `Params a` is just a newtype over `a -> Something`, but I don't see how to do it using the public interface.

It may also be nice to document how to use the `Divisible` and `Decidable` interfaces (e.g. by having examples of their use in the haddocks).

**Comment by rpglover64** [comment]
Another related use case is `(a -> Params ()) -> Params a`.

**Comment by nikita-volkov** [comment]
One way of easily wrapping your head around those interfaces is thinking of `Contravariant` as a dual of `Functor`, `Divisible` - `Applicative`, `Decidable` - `Alternative`.

Possibly I'm wrong but I believe that what you're looking for is a dual of `Monad`, but the API does not support any such interface. What do you need that for anyway? Can you please provide a specific use-case?

**Comment by rpglover64** [comment]
> One way of easily wrapping your head around those interfaces is thinking of [...] as a dual of [...]

Good intuition for duality is hard to come by, and a few examples will make the library more usable with less study. I also disagree that this ways is "easy".

I have two use-cases in mind:

* I have a function

  ```
enc1 :: Foo -> Params (Bar) -- Bar is a record type with fields a, b, and c
enc1 x =
  contramap a (value int8) <>
  contramap (const x) (value int8) <>
  contramap b (value int8) <>
  contramap c (value int8)
```
  I find this formulation easier to read and write and work with than
  ```
enc2 :: Params (Foo, Bar)
enc2 =
  contramap (a . snd) (value int8) <>
  contramap fst (value int8) <>
  contramap (b . snd) (value int8) <>
  contramap (c . snd) (value int8)
```
  or
  ```
enc3 :: Params (Foo, Bar)
enc3 =
  contramap (a . snd) (value int8) <>
  contramap fst (value int8) <>
  contramap snd (
    contramap b (value int8) <>
    contramap c (value int8)
  )
```
  but sometimes I need something of the type of `enc2`, so I'd rather write `enc1` and have a function that converts it to `enc2`

* I have a sum type
  ```
SumThing = A Int | B Int | C Int
```
  and I would like to build a `Params SumThing`. The straightforward approach of
  ```
encSum x = case x of
  A i -> fromIntegral i >$ value int8
  ...
```
  has the wrong type, and while in this example I could extract the value first, that is not always the case.  `Divisible` makes it possible, but only by manually decomposing the `SumThing` into `Either Int (Either Int Int)`, which is much less pleasant to work with than a case statement, especially as the number of constructors increases.

I'm pretty sure that I am not looking for the contravariant version of monad (though that might solve the problem as well); it feels more related to CCCs because I'm thinking of it as currying. I'm also pretty sure that this is possible with `Params`, but not with an arbitrary `Divisible`.

Consider the simpler case of `Predicate` (I believe that it's fair to analogize `Params` with `Predicate`, but correct me if I'm wrong):  I want a function that takes an `a -> Predicate b` (which is newtype-equivalent to `a -> b -> Bool`) to `Predicate (a,b)` (which is newtype-equivalent to `(a,b) -> Bool`).

**Comment by nikita-volkov** [comment]
So basically your problem is that you find the contravariant APIs inconvenient. However Hasql's API uses them for plenty valid reasons and trying to sidestep them won't do you any good.

For instance, you're missing the fact that `Params` is a more involved thing than `Predicate` or whatever other function wrapper. For one thing, atop of encoding data it encodes a reference to a positional placeholder. So a case expression simply can't work here, because it will miss the information about which placeholder you're referring to on each branch. That's one.

From conceptual perspective there's a problem with what you want. Encoders are really meant to be a static thing. Think of them as of specifications of how to encode data, not the data itself. They are meant to be isolated from data. Declaring a function like `enc1 :: Foo -> Params (Bar)` means that you're parameterising the specification itself, instead of providing a parameter to the encoder as you expect. And that really does not make much sense, since the SQL statement, that you'll bind the encoder to will most probably be static itself. If it won't then you'll lose all the benefits of Prepared Statements.

So I'm afraid there's no other way than just getting used to the contravariant APIs.
Sidestepping them won't do you any good and will only bring you problems. There are ways to evolve the contravariant APIs though, I'm pretty sure.

**Comment by tomjaguarpaw** [comment]
@rpglover64 Are you still having problems implementing these?  I can give you the answers if so.  Just tell me which specific type signatures you want implemented.

**Comment by rpglover64** [comment]
Thank you.

I'm not currently touching the hasql part of the code base I'm working on,
but when I return to it, `uncurry :: (a -> Params b) -> Params (a, b)` will
be useful if it is possible.

**Comment by tomjaguarpaw** [comment]
@rpglover64 Actually I'm pretty sure this is not possible in general, sorry.

**Comment by neongreen** [comment]
I would also love to have a `(a -> Params ()) -> Params a`.

My usecase is that I want the compiler to warn me when I add a new field to the data type and forget to update the row encoder. So, I'd like to write something like this:

```haskell
-- collapse :: (a -> Params ()) -> Params a

data FooRow = FooRow A B C

fooRow :: Params FooRow
fooRow = collapse $ \FooRow a b c ->
    a >$ param encoderA <>
    b >$ param encoderB <>
    c >$ param encoderC
```

**Comment by neongreen** [comment]
Or actually it could even be `collapse :: (a -> Params Void) -> Params a`, perhaps.

**Comment by nikita-volkov** [comment]
@neongreen 

Such operation is indeed possible to implement in the current version of Hasql, but I doubt it will be in future versions. So I would rather not motivate the users to evolve usage patterns, which have a risk of not surviving.

Such operation is also overly powerful for your case. What you want can already be achieved with combination of `contramap` and `contrazip`. The overpower is that it allows to provide different encoders based on input, which basically is nonsense and something that I'd rather prohibit with the type system.

But the issue is actually more of idiomatical here. It's not the Haskell Model that must determine your statements (and hence encoders), it is the database schema. If you think this way, then your model can only change when the need arises from the codecs, not the other way.

**Comment by neongreen** [comment]
> If you think this way, then your model can only change when the need arises from the codecs, not the other way.

Sure, but my code still becomes more fragile as the result. It's hard to forget to update the row type when the table schema changes, but it's easier to forget to update the row type encoder.

This said, I can kinda solve this issue with HKD (or by using hasql-class) so I am not blocked by this.

**Comment by mitchellwrosen** [comment]
@neongreen Could you please share your technique for solving this issue? I'm interested.

**Comment by neongreen** [comment]
@mitchellwrosen 

The `hasql-class` library includes `Generic`-based encoders and decoders for records IIRC, which solves the problem.

To solve it with [HKD](https://reasonablypolymorphic.com/blog/higher-kinded-data/), you have two options.

Option A: rewrite your type HKD-style:

```haskell
data FooRow f = FooRow
    { a :: f Int
    , b :: f Bool
    , c :: f Text
    }
```

Now you can construct a `FooRow` of encoders:

```haskell
fooRow :: FooRow (Const (Params FooRow))
fooRow = FooRow
    { a = a >$< param int64
    , b = b >$< param bool
    , c = c >$< param text
    }
```

And turn it into an encoder for `FooRow`:

```haskell
import Data.Barbies

fooRowParams :: Params (FooRow Identity)
fooRowParams = bsequence' fooRow
```

Note that this is still not an ideal solution – in an ideal world you wouldn't have to write `a = a >$< param ...`, you'd write just `a = param ...`. But at least the compiler checks that I have gone through all fields.

Option B: if you don't want HKD-style types, use `higgledy`: http://hackage.haskell.org/package/higgledy-0.3.0.0/docs/Data-Generic-HKD-Build.html. It's slightly worse because you won't be able to supply field names when constructing a row of encoders. This will get solved if https://github.com/i-am-tom/higgledy/issues/15 goes through.

**Comment by neongreen** [comment]
I'm leaning heavily towards using generic encoders in this case.
